### PR TITLE
Fix momentum bug in CyclicLR

### PR DESCRIPTION
**Pull request by emerali**
Resolves issue #19003

The author of this issue also asked that `cycle_momentum` default to `False` if the optimizer does not have a momentum parameter, but I'm not sure what the best way to do this would be. Silently changing the value based on the optimizer may confuse the user in some cases (say the user explicitly set `cycle_momentum=True` but doesn't know that the Adam optimizer doesn't use momentum). 

Maybe printing a warning when switching this argument's value would suffice?

### CODE REVIEW

**Comment by soumith** [comment]
@emerali this looks good, thanks for fixing the bug. Can you add a test to test/test_optim.py that triggers this case so that we dont regress in the future?

File: test/test_optim.py [hunk 1]
```diff
@@ -961,6 +961,15 @@ def test_cycle_lr_exp_range_mode_step_size_up_down(self):
                              mode='exp_range', gamma=gamma)
         self._test_cycle_lr(scheduler, lr_targets, momentum_targets, len(lr_target))
 
+    def test_cycle_lr_with_momentumless_optimizer(self):
+        adam_opt = optim.Adam(self.net.parameters())
+        scheduler = CyclicLR(adam_opt, cycle_momentum=False)
```
Comment (ezyang) [inline test/test_optim.py]: You need to actually call `self._test_cycle_lr` here...

**Review by ezyang — CHANGES_REQUESTED** [review]
test is incomplete

**Comment by emerali** [comment]
I've fixed the test as requested. It required a little bit of a hack though, as the `TestLRScheduler._test_cycle_lr` function uses the object's internal optimizer (SGD), so I temporarily swapped it out with an Adam optimizer object inside of the test body.

**Comment by ezyang** [comment]
> It required a little bit of a hack though, as the TestLRScheduler._test_cycle_lr function uses the object's internal optimizer (SGD), so I temporarily swapped it out with an Adam optimizer object inside of the test body.

Would you mind elaborating on this more? (Or @soumith, can you tell what's going on here?) I'm not sure why use of the internal optimizer causes a problem for the test (I might be missing something obvious!)

**Comment by emerali** [comment]
The `TestLRScheduler` object carries around and SGD optimizer to avoid having to instantiate one for every test. This gets in the way for our very specific case in which we need to use Adam (or really any optimizer that doesn't use momentum) in order to test that the momentum bug in CyclicLR is fixed (the bug is described in more detail in the corresponding issue).

Does that answer your question?

**Comment by emerali** [comment]
Friendly nudge as this seems to have fallen under the radar.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
I added your comment here as a comment in the code. This is good to go as soon as the rebased CI passes.

**Comment by mugheesahmad** [comment]
Sorry for the immature question but I am still facing this issue at my PyTorch. kindly suggest me how can I update this in my PyTorch framework.

**Comment by zwep** [comment]
Yup, same here. Still the same error when using CyclicLR and Adam together

**Comment by ezyang** [comment]
Try upgrading your version of PyTorch; the latest version should have this fix.

**Comment by zwep** [comment]
Yeah, even the 1.4 version has this fix. (Just so you know 1.5 is not your only option. Since in 1.5 you also need cuda version 10.2 or so to have it working)

**Comment by ytfksw** [comment]
This issue has not yet been fixed in the version 1.6...

**Comment by austinmw** [comment]
I'm using 1.6 and encountered this bug

**Pull request closed without merging**